src/internal/data/metrics.go

```diff
@@ -15,3 +22,73 @@ func (cs CollectionStats) IsZero() bool {
 func (cs CollectionStats) String() string {
 	return cs.Details
 }
+
+type KindStats struct {
+	BytesRead     int64
+	ResourceCount int64
+}
+
+type ExportStats struct {
+	// data is kept private so that we can enforce atomic int updates
+	// TODO(meain): Should we be exposing details.ItemType?
+	data map[details.ItemType]KindStats
+}
+
+func (es *ExportStats) UpdateBytes(kind details.ItemType, bytesRead int64) {
+	if es.data == nil {
+		es.data = map[details.ItemType]KindStats{}
+	}
+
+	ks := es.data[kind]
+	atomic.AddInt64(&ks.BytesRead, bytesRead)
+	es.data[kind] = ks
+}
+
+func (es *ExportStats) UpdateResourceCount(kind details.ItemType) {
+	if es.data == nil {
+		es.data = map[details.ItemType]KindStats{}
+	}
+
+	ks := es.data[kind]
+	atomic.AddInt64(&ks.ResourceCount, 1)
+	es.data[kind] = ks
+}
+
+func (es *ExportStats) GetStats() map[details.ItemType]KindStats {
+	return es.data
+}
+
+type statsReader struct {
+	reader io.ReadCloser
+	kind   details.ItemType
+	stats  *ExportStats
+}
+
+func (sr *statsReader) Read(p []byte) (n int, err error) {
```

Comment:
nit: this one is purely preference on my part, but can we keep explicit returns of data instead of named return variables. It takes my mind a few extra seconds to go through the logic of empty return statement -> something that still returns values -> using named return values 😅